### DISCUSSION TITLE
webpackHotUpdate  currentUpdate is undefined

### DISCUSSION THREAD

**Discussion by freshhan163** [webpack/webpack#14873] [General]
# Bug report

**What is the current behavior?**

when modify app.vue, then refresh page, there will be an error: 

```js
app.js:939 Uncaught TypeError: Cannot set properties of undefined (setting './node_modules/babel-loader/lib/index.js??clonedRuleSet-1.use!./node_modules/vue-loader/dist/templateLoader.js??ruleSet[1].rules[3]!./node_modules/vue-loader/dist/index.js??ruleSet[0].use[0]!./src/App.vue?vue&type=template&id=7ba5bd90')
    at self.webpackHotUpdatepc_live_next
```

after check, it is currentUpdate is undefined:
```js
/******/           self["webpackHotUpdatepc_live_next"] = (chunkId, moreModules, runtime) => {
/******/ 			for(var moduleId in moreModules) {
/******/ 				if(__webpack_require__.o(moreModules, moduleId)) {
/******/ 					currentUpdate[moduleId] = moreModules[moduleId];
/******/ 					if(currentUpdatedModulesList) currentUpdatedModulesList.push(moduleId);
/******/ 				}
/******/ 			}
/******/ 			if(runtime) currentUpdateRuntime.push(runtime);
/******/ 			if(waitingUpdateResolves[chunkId]) {
/******/ 				waitingUpdateResolves[chunkId]();
/******/ 				waitingUpdateResolves[chunkId] = undefined;
/******/ 			}
/******/ 		};
```

And to compare with webpack4, it is 
```js
/******/ 	function hotAddUpdateChunk(chunkId, moreModules) {
/******/ 		if (!hotAvailableFilesMap[chunkId] || !hotRequestedFilesMap[chunkId])
/******/ 			return;
/******/ 		hotRequestedFilesMap[chunkId] = false;
/******/ 		for (var moduleId in moreModules) {
/******/ 			if (Object.prototype.hasOwnProperty.call(moreModules, moduleId)) {
/******/ 				hotUpdate[moduleId] = moreModules[moduleId];
/******/ 			}
/******/ 		}
/******/ 		if (--hotWaitingFiles === 0 && hotChunksLoading === 0) {
/******/ 			hotUpdateDownloaded();
/******/ 		}
/******/ 	}
```

**If the current behavior is a bug, please provide the steps to reproduce.**

**What is the expected behavior?**
I would expect that when currentUpdate is undfined, return.

**Other relevant information:**
webpack version:  5.64.3
Node.js version: v12.19.0
Operating System:  OS
Additional tools: vue3.0

**Comment by alexander-akait** [comment]
Please create reproducible test repo

**Reply by freshhan163** [reply]
repo is: https://github.com/freshhan163/test-vue3-ssr

how to recurrent：
start project, then modify app.vue, then update pages, you will see this bug.

**Reply by alexander-akait** [reply]
You have broken peer deps:
```
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! While resolving: test-vue3-ssr@0.0.0
npm ERR! Found: vue-router@4.0.12
npm ERR! node_modules/vue-router
npm ERR!   vue-router@"^4.0.12" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer vue-router@"^3.0.0" from vuex-router-sync@5.0.0
npm ERR! node_modules/vuex-router-sync
npm ERR!   vuex-router-sync@"^5.0.0" from the root project
```

Anyway I can't reproduce your problem, but I don't use your lock files because you have packages located on `npm.corp.kuaishou.com` (not available to me), so all deps were updated (i.e. I used `rm -rf package-lock.json && npm i`) and no problems.

**Reply by freshhan163** [reply]
when first run project, it is no problem.

but after you modify src/App.vue , save and refresh page, you will find error on chrome console.

![image](https://user-images.githubusercontent.com/18698094/144407921-36266c84-16f9-46e7-9d94-33e759a55a87.png)

please help check it again. 
Thanks!

**Reply by alexander-akait** [reply]
How many web socket do you have on the page (look at dev tools)

**Reply by freshhan163** [reply]
I only found one.
<img width="888" alt="截屏2021-12-03 上午10 49 13" src="https://user-images.githubusercontent.com/18698094/144536571-2e12e69d-e985-45b5-b59f-af2d423963a3.png">
<img width="898" alt="截屏2021-12-03 上午10 49 01" src="https://user-images.githubusercontent.com/18698094/144536589-588905d7-ae27-4303-8c1e-489d994c43cc.png">

**Reply by alexander-akait** [reply]
Based on the problem, bug in `vue-loader`... Do you use official docs to setup it?

**Reply by freshhan163** [reply]
you can see webpack config in repo: https://github.com/freshhan163/test-vue3-ssr

I think it is not `vue-loader` problem. 

can you reproduce my problem?

The Problem is in HotModuleReplacementPlugin, It seems like didn't consider SSR condition. After I modify App.vue, then refresh page,  this errror will appear.

**Reply by alexander-akait** [reply]
Sorry, I don't see any problems with `HotModuleReplacementPlugin`, vue-loader adds additional logic for SSR and looks it is broken

**Reply by freshhan163** [reply]
@vankop  Is there a solution for this problem?

**Reply by utkanpar** [reply]
@freshhan163 I had a similar issue with <https://github.com/webpack/webpack/discussions/14873#discussioncomment-2283858> but with a different error. The issue in my scenario was XXX.hot-update.js was by mistake loaded into the client side as a standalone script. The hot-update.js script is not supposed to be loaded on the client side. More info here:  https://github.com/webpack/webpack/issues/11670#issuecomment-715512003

**Comment by lihancong** [comment]
try to remove the globalObject: 'this' setting from output

**Reply by pisetski** [reply]
Hey @lihancong, your solution works for me. Thanks!

**Comment by utkanpar** [comment]
Hello @alexander-akait @lihancong ,  we are seeing similar error in our react project as well after I modify my code in .tsx file and after HMR reloads the page.

We are not using globalObject: 'this' setting in the output object.

Please check for the error below

<img width="618" alt="Screenshot 2022-03-02 173849" src="https://user-images.githubusercontent.com/69013206/156462080-21677dbd-18ad-49aa-9aac-911db379317d.png">

The error is coming from jsonp chunk loading:

<img width="530" alt="error2" src="https://user-images.githubusercontent.com/69013206/156462144-a456c60e-0c41-415d-9825-f38b56e0ac0c.png">

)


Looks like variable currentUpdateRuntime, currentUpdate are both undefined.


Other relevant information:
webpack version: 5.69.1
Node.js version: v12.22.1
Operating System: Windows

**Reply by vankop** [reply]
please create reproducible repo

**Reply by utkanpar** [reply]
Hello @vankop , Apologies for the late response. 

Here is the reproducible repo with the issue: https://github.com/utkanpar/Webpack-HMR-properties-undefined-error

I have also attached the webpack-config file and my yarn.lock file at the root level of the repo

**Reply by utkanpar** [reply]
Hi @vankop / @alexander-akait, can you please help taking a look into the issue? As requested, I have attached the reproducible repo with the issue: https://github.com/utkanpar/Webpack-HMR-properties-undefined-error

**Reply by vankop** [reply]
there was already several reports. there is an issue with `globalObject: "this"` since in strict mode `this` is undefined for function expression and we can't remove strict mode in some cases..

**Reply by vankop** [reply]
please use another globalObject if this is an issue for you

**Reply by utkanpar** [reply]
Hi @vankop, We are not using `globalObject: "this"` but are defaulting to globalObject to `self`

If you take a look at the webpack config file in the attached repro repo: https://github.com/utkanpar/Webpack-HMR-properties-undefined-error/blob/main/webpack-config.js and search for `config.ouput`, you'll see that we are not setting `globalObject` explicitly.

![image](https://user-images.githubusercontent.com/69013206/162339675-c9de73a3-a19a-4404-a3c6-64f269f78468.png)

The issue is repro even with `globalObject: "self"`

**Reply by vankop** [reply]
your repro need access to `https://microsoft.pkgs.visualstudio.com/_packaging/*`. Please try to reduce it.. ( also small webpack config would be easier to understand )

**Reply by utkanpar** [reply]
Hi @vankop, I have fixed the access issue. You should be able to pull in the packages from the feed now. Please let me know if you are still having access issues.

Also, I have tried cleaning up the webpack config file and making it smaller : https://github.com/utkanpar/Webpack-HMR-properties-undefined-error/blob/main/webpack-config.js

**Reply by utkanpar** [reply]
Hi @vankop, checking to see if you were able to pull in the packages from the feed? I fixed the access issue in the repro repo. Also cleaned up the webpack config to a smaller file.

**Reply by vankop** [reply]
this does not work for me... maybe because Im using mac..

**Reply by utkanpar** [reply]
Hi @vankop, what is the error that you are seeing? Can you share a screenshot maybe?

I am successfully able to do yarn install and also start the server and run the app on my Mac.

**Reply by vankop** [reply]
node sass does not work with M1 I think
```
In file included from /Users/ivankopeykin/.node-gyp/16.13.2/include/node/node.h:63:
In file included from /Users/ivankopeykin/.node-gyp/16.13.2/include/node/v8.h:30:
/Users/ivankopeykin/.node-gyp/16.13.2/include/node/v8-internal.h:492:38: error: no template named 'remove_cv_t' in namespace 'std'; did you mean 'remove_cv'?
            !std::is_same<Data, std::remove_cv_t<T>>::value>::Perform(data);
                                ~~~~~^~~~~~~~~~~
                                     remove_cv
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/type_traits:710:50: note: 'remove_cv' declared here
template <class _Tp> struct _LIBCPP_TEMPLATE_VIS remove_cv
                                                 ^
1 error generated.
make: *** [Release/obj.target/binding/src/binding.o] Error 1
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/Users/ivankopeykin/Repositories/Webpack-HMR-properties-undefined-error/node_modules/@msdyn365-commerce/theming-internal/node_modules/node-gyp/lib/build.js:262:23)
gyp ERR! stack     at ChildProcess.emit (node:events:390:28)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:290:12)
gyp ERR! System Darwin 21.3.0
gyp ERR! command "/Users/ivankopeykin/.nvm/versions/node/v16.13.2/bin/node" "/Users/ivankopeykin/Repositories/Webpack-HMR-properties-undefined-error/node_modules/@msdyn365-commerce/theming-internal/node_modules/node-gyp/bin/node-gyp.js" "rebuild" "--verbose" "--libsass_ext=" "--libsass_cflags=" "--libsass_ldflags=" "--libsass_library="
gyp ERR! cwd /Users/ivankopeykin/Repositories/Webpack-HMR-properties-undefined-error/node_modules/@msdyn365-commerce/theming-internal/node_modules/node-sass
gyp ERR! node -v v16.13.2
```

**Reply by utkanpar** [reply]
@vankop What is the node version you are using on your local machine? The app currently only support node version <= 12 lts

<img width="352" alt="image" src="https://user-images.githubusercontent.com/69013206/163319725-ebd90355-6f71-43d6-9dd7-f8e5641d658b.png">

Can you please run the app using node 12 and see if you still see the issue with yarn install?

Also, since you are using Mac, in package.json you might want to replace "SET" with 'set' (lowercase).

**Reply by vankop** [reply]
```bash
$ yarn start                                                                                                                                                        [14:02:21]
yarn run v1.22.18
$ SET NODE_OPTIONS=--max_old_space_size=4096 && yarn msdyn365b start local --use-eslint
/bin/sh: SET: command not found
error Command failed with exit code 127.
```

**Reply by utkanpar** [reply]
Hi @vankop , the package was developed in windows, since you are using Mac, in the package.json you will want to replace "SET" with 'set' (lowercase) or you can pull in the latest from the repo branch (https://github.com/utkanpar/Webpack-HMR-properties-undefined-error). I have updated the package.json there to work on Mac.

The reason you see that error is because:
Since the package was developed in Windows and Windows isn't case-sensitive, but bash is.

**Reply by utkanpar** [reply]
Hi @vankop , checking to see if you were able to run the repro repo?

**Reply by utkanpar** [reply]
Hi @vankop , pinging again to check if you were able to run the repo?

**Reply by utkanpar** [reply]
Hello @vankop / @alexander-akait, are you running into any issues running the repo?

**Comment by f3ndot** [comment]
Hi @vankop @alexander-akait I've come across this whilst learning about Webpack, going through the guide near verbatim. I've created a reproducible repo here: https://github.com/f3ndot/webpack-currentupdate-undefined

Let me know if you need more information or to try stuff in the repo

**Comment by vankop** [comment]
@f3ndot 
your html file (`index.html`) content looks like 
```html
<html>
  <head>
    <meta charset="utf-8">
    <title>TypeScript Playground</title>
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <script defer src="index.bundle.js"></script><script defer src="print.bundle.js"></script>
</head>
  <body>
  </body>
</html>
```

**Note** both entrypoints are presented in html file and that is wrong.. You can setup html plugin to create html file per entrypoint

**Reply by vankop** [reply]
not sure but maybe `dependOn` option could help here https://webpack.js.org/configuration/entry-context/#entry-descriptor

UPD: Yeah, this setting helps.

**Reply by f3ndot** [reply]
Hey @vankop I'm a little confused by your response, since I was following the Webpack 5 Guide near verbatim:

1. That .html file was generated by `HtmlWebpackPlugin`
2. The [guide indicated](https://webpack.js.org/guides/output-management/#setting-up-htmlwebpackplugin) to specify an `entrypoint: {}` object with both an `index` and `print` property/entry whilst using the webpack plugin. Seems to suggest the expected result is a single .html referencing the two entrypoints.
3. The [next page in the guide](https://webpack.js.org/guides/development/#using-webpack-dev-server) instructs how to setup the dev server/use watch mode, where the config.js shown begins to match mine more or less. It was around this point I experienced the issues.

So did I make a mistake following the guide, or is the guide incorrect for how one is supposed to use these concepts. Does the guide need to be updated to reflect what's necessary to make this work? More broadly: am I to understand it's incorrect to just have 2+ entrypoints be used in a single .html file for hot update?

**Reply by f3ndot** [reply]
Ah I see. Reading further, I see this passage on the [Code Splitting's Entry dependencies](https://webpack.js.org/guides/code-splitting/#entry-dependencies) section of the guide:

> If we're going to use multiple entry points on a single HTML page, optimization.runtimeChunk: 'single' is needed too, otherwise we could get into trouble described [here](https://bundlers.tooling.report/code-splitting/multi-entry/).

So to me this seems like a small update is necessary to the prior chapter/section of the guide ([Development](https://webpack.js.org/guides/development/)) to state the same thing so if someone uses the dev server/webpack watch for hot reloads, it won't bug out. That is, explicitly, have the webpack.config.js in the prior chapter have the `{ optimization: { runtimeChunk: 'single' }` added. In other words, following the guide verbatim start to finish, people will experience what I did until they ignore and move on to the code splitting section.

**Reply by f3ndot** [reply]
To help future noobs not get tripped up by this like I did, submitted a PR against the Development section of the guide: https://github.com/webpack/webpack.js.org/pull/6098

**Comment by nico-ledesma** [comment]
I stumbled across this problem too, not a typical setup, but just in case it helps someone else, this is what happened:

- HTML was generated an Express server.
- WebpackDevServer was serving assets, and proxying everything else to the Express server.
- When bundling, I was creating a custom manifest with `WebpackManifestPlugin`.
- Express server would read this manifest and print relevant `<script>` tags in HTML.

It turns out that `[hash].hot-update.js` scripts were being written to the manifest, and output to the HTML file, causing that error.

Filtering those files fixed everything :v:

**Reply by vankop** [reply]
remove hmr plugin in production mode

**Reply by rickgladwin** [reply]
> It turns out that `[hash].hot-update.js` scripts were being written to the manifest, and output to the HTML file, causing that error.
> 
> Filtering those files fixed everything ✌️

If you can recall, where did you set up the filter? In the main webpack config itself? Or in a WebpackDevServer config? 
Thanks!

**Reply by nico-ledesma** [reply]
I was using a [custom generator function](https://github.com/shellscape/webpack-manifest-plugin#generate), and did something like this:

```js
const manifestGenerator = ( seed, files, entries ) => {
	const manifest = {};

	// ...

	Object.keys( entries ).forEach( entryPoint => {
		const entryPointFiles = entries[ entryPoint ];
		const jsAssetRegexp = /(?<!hot-update).js$/;
		entryPointFiles.forEach( fileName => {
			if( jsAssetRegexp.test( fileName ) ) {
				// These are the JS files you want in your manifest
			}
		} );
	} );

	// ...

	return manifest;
}
```

Hope it helps!

**Reply by rickgladwin** [reply]
🙏🏻 !!

**Comment by skrat** [comment]
I am hitting this error. I have multiple entry points, and very simple service worker entry point. But using `runtimeChunk: 'single'` results in my service worker not working. I've no idea how. Here's the worker, the fetch handler is never fired with `runtimeChunk: 'single'`, but works fine without it, but then HMR breaks:

```
const CARDS = CARDS_DATA;

const listing = (prefix) =>
  Object.keys(CARDS)
    .filter(p => p.startsWith(prefix))
    .map(p => p.substring(prefix.length))
    .sort().join("\n")

const PATTERNS = [
  ["/cards/user", { body: listing("/user/") }],
  ...Object.entries(CARDS).map(([k, v]) => (
    ["/cards" + k, { body: v }]
  ))
];

self.addEventListener("install", () => {
  console.log("[SW] Service worker installed");
  self.skipWaiting();
});

self.addEventListener("activate", (event) => {
  console.log("[SW] Service worker activated");
  event.waitUntil(clients.claim());
});

self.addEventListener("fetch", (event) => {
  const url = new URL(event.request.url);
  for (const [p, val] of PATTERNS) {
    if (p === url.pathname) {
      console.log(`[SW] fetch ${event.request.method} - ${url.pathname}`);
      if (event.request.method === "GET") {
        event.respondWith(new Response(val.body));
      } else if (event.request.method === "POST") {
        (async () => {
          val.body = await event.request.text();
        })();
        event.respondWith(new Response())
      }
      break;
    }
  }
});
```
